File: Leetcode/739-Daily-Temperatures/739-Daily-Temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> result(temperatures.size());
        stack<int> index;

        for (int i = 0; i < temperatures.size(); i++) {
            while(!index.empty() && temperatures[i] > temperatures[index.top()]) {
                result[index.top()] = i - index.top();
                index.pop();
            }

            index.push(i);
        }

        return result;
    }
};

File: Leetcode/739-Daily-Temperatures/739-Daily-Temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] result = new int[temperatures.length];
        MonotonicStack stack = new MonotonicStack();

        for (int i = temperatures.length - 1; i >= 0; i--) {
            result[i] = stack.daysUntilGreaterTemperature(temperatures[i], i);
        }

        return result;
    }

    private class MonotonicStack {
        Deque<int[]> stack; // Temperature, day 

        public MonotonicStack() {
            stack = new ArrayDeque<>();
        }

        public int daysUntilGreaterTemperature(int temperature, int day) {
            int[] currGreaterTemperature, currTemperature = {temperature, day};

            while (!stack.isEmpty()) {
                currGreaterTemperature = stack.peek();

                if (currGreaterTemperature[0] > temperature) {
                    
                    stack.push(currTemperature);

                    return currGreaterTemperature[1] - day;
                }

                stack.pop();
            }

            stack.push(currTemperature);

            return 0;
        }
    }
}